File: android/app/src/main/java/tech/mrzeapple/ciphercodex/sync/KosyncClient.kt
package tech.mrzeapple.ciphercodex.sync

import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.withContext
import kotlinx.serialization.json.Json
import kotlinx.serialization.json.JsonObject
import kotlinx.serialization.json.JsonPrimitive
import kotlinx.serialization.json.buildJsonObject
import kotlinx.serialization.json.contentOrNull
import kotlinx.serialization.json.doubleOrNull
import kotlinx.serialization.json.intOrNull
import kotlinx.serialization.json.jsonObject
import kotlinx.serialization.json.longOrNull
import kotlinx.serialization.json.put
import okhttp3.HttpUrl
import okhttp3.HttpUrl.Companion.toHttpUrlOrNull
import okhttp3.MediaType.Companion.toMediaType
import okhttp3.OkHttpClient
import okhttp3.Request
import okhttp3.RequestBody
import okhttp3.RequestBody.Companion.toRequestBody
import okhttp3.Response
import java.io.IOException
import java.util.concurrent.TimeUnit
import kotlin.math.floor

/** kosync protocol client. Wire format and error codes follow KOReader and the
 *  reference koreader-sync-server; shared types live in KosyncContracts. */
class KosyncClient : KosyncApi {

    private val client: OkHttpClient by lazy {
        OkHttpClient.Builder()
            .connectTimeout(10, TimeUnit.SECONDS)
            .callTimeout(30, TimeUnit.SECONDS)
            .build()
    }

    private val json = Json { ignoreUnknownKeys = true }

    override suspend fun register(account: KosyncAccount): KosyncResult<Unit> = withContext(Dispatchers.IO) {
        val url = url(account, "users", "create") ?: return@withContext badUrl(account)
        val body = buildJsonObject {
            put("username", account.username)
            put("password", account.userKey)
        }
        // Registration is the one call that must NOT carry auth headers.
        execute({ request(url).post(body.toBody()).build() }) { response ->
            if (response.code == 201) KosyncResult.Ok(Unit) else errorFrom(response)
        }
    }

    override suspend fun authorize(account: KosyncAccount): KosyncResult<Unit> = withContext(Dispatchers.IO) {
        val url = url(account, "users", "auth") ?: return@withContext badUrl(account)
        execute({ request(url).auth(account).get().build() }) { response ->
            if (response.code == 200) KosyncResult.Ok(Unit) else errorFrom(response)
        }
    }

    override suspend fun getProgress(account: KosyncAccount, document: String): KosyncResult<RemoteProgress?> =
        withContext(Dispatchers.IO) {
            val url = url(account, "syncs", "progress", document) ?: return@withContext badUrl(account)
            execute({ request(url).auth(account).get().build() }) { response ->
                if (response.code != 200) return@execute errorFrom(response)
                val obj = runCatching {
                    json.parseToJsonElement(response.body?.string().orEmpty()).jsonObject
                }.getOrNull()
                    ?: return@execute KosyncResult.Err(response.code, "Unexpected response from server")
                // "No record" is a 200 with an empty object: detect by the absent
                // percentage field, never by status.
                val percentage = obj.number("percentage")?.toFloat()
                    ?: return@execute KosyncResult.Ok(null)
                KosyncResult.Ok(
                    RemoteProgress(
                        document = obj.string("document") ?: document,
                        progress = obj.string("progress").orEmpty(),
                        percentage = percentage,
                        device = obj.string("device"),
                        deviceId = obj.string("device_id"),
                        timestamp = (obj["timestamp"] as? JsonPrimitive)?.longOrNull,
                    )
                )
            }
        }

    override suspend fun updateProgress(account: KosyncAccount, progress: RemoteProgress): KosyncResult<Unit> =
        withContext(Dispatchers.IO) {
            val url = url(account, "syncs", "progress") ?: return@withContext badUrl(account)
            val body = buildJsonObject {
                put("document", progress.document)
                put("progress", progress.progress)
                // KOReader convention: truncate (never round up) to 4 decimals.
                put("percentage", floor(progress.percentage.toDouble() * 10000.0) / 10000.0)
                progress.device?.let { put("device", it) }
                progress.deviceId?.let { put("device_id", it) }
            }
            execute({ request(url).auth(account).put(body.toBody()).build() }) { response ->
                if (response.code == 200) KosyncResult.Ok(Unit) else errorFrom(response)
            }
        }

    /** Every kosync request needs this accept header — the reference server
     *  responds 412 without it. */
    private fun request(url: HttpUrl): Request.Builder =
        Request.Builder().url(url).header("accept", "application/vnd.koreader.v1+json")

    private fun Request.Builder.auth(account: KosyncAccount): Request.Builder = this
        .header("x-auth-user", account.username)
        // Servers compare the key verbatim; the md5 hex must be lowercase.
        .header("x-auth-key", account.userKey.lowercase())

    private fun url(account: KosyncAccount, vararg segments: String): HttpUrl? {
        // Stored without a trailing slash, but defend anyway.
        val base = account.serverUrl.trim().trimEnd('/').toHttpUrlOrNull() ?: return null
        return base.newBuilder().apply {
            segments.forEach { addPathSegment(it) } // percent-encodes each segment
        }.build()
    }

    private fun badUrl(account: KosyncAccount): KosyncResult.Err =
        KosyncResult.Err(null, "Invalid server URL: ${account.serverUrl}")

    /** Takes a factory, not a Request: OkHttp throws IllegalArgumentException for
     *  non-ASCII header values (e.g. auth headers from a non-ASCII username) at
     *  build time, so construction must happen inside the try to be catchable. */
    private fun <T> execute(buildRequest: () -> Request, handle: (Response) -> KosyncResult<T>): KosyncResult<T> =
        try {
            client.newCall(buildRequest()).execute().use(handle)
        } catch (e: IOException) {
            KosyncResult.Err(null, e.message ?: e.javaClass.simpleName)
        } catch (e: IllegalArgumentException) {
            KosyncResult.Err(null, e.message ?: "Invalid credentials or URL")
        }

    private fun errorFrom(response: Response): KosyncResult.Err {
        val obj = runCatching {
            json.parseToJsonElement(response.body?.string().orEmpty()).jsonObject
        }.getOrNull()
        val code = (obj?.get("code") as? JsonPrimitive)?.intOrNull
        val message = friendlyMessage(code)
            ?: obj?.string("message")
            ?: response.message.ifBlank { "HTTP ${response.code}" }
        return KosyncResult.Err(response.code, message)
    }

    /** Known koreader-sync-server error codes, mapped to readable text. */
    private fun friendlyMessage(code: Int?): String? = when (code) {
        2001 -> "Unauthorized"
        2002 -> "Username is already registered"
        2003 -> "Invalid request"
        2005 -> "Registration is disabled on this server"
        else -> null
    }

    private fun JsonObject.toBody(): RequestBody = toString().toRequestBody(MEDIA_JSON)

    private fun JsonObject.string(key: String): String? = (this[key] as? JsonPrimitive)?.contentOrNull

    /** Lenient: the value may arrive as an int, float, or numeric string. */
    private fun JsonObject.number(key: String): Double? = (this[key] as? JsonPrimitive)?.doubleOrNull

    private companion object {
        val MEDIA_JSON = "application/json".toMediaType()
    }
}

File: remarkable2-os/src/sync/kosyncclient.cpp
#include "kosync.h"

#include <QEventLoop>
#include <QJsonDocument>
#include <QJsonObject>
#include <QNetworkAccessManager>
#include <QNetworkReply>
#include <QNetworkRequest>

namespace ccx::kosync {

namespace {

const char *const kAccept = "application/vnd.koreader.v1+json";

// One raw HTTP outcome: httpCode == -1 means no response (transport failure).
struct Raw {
    int httpCode = -1;
    QByteArray body;
    QString transportError;
};

// Known koreader-sync-server error codes, mapped to readable text (matches Android).
QString friendlyMessage(int code)
{
    switch (code) {
    case 2001: return QStringLiteral("Unauthorized");
    case 2002: return QStringLiteral("Username is already registered");
    case 2003: return QStringLiteral("Invalid request");
    case 2005: return QStringLiteral("Registration is disabled on this server");
    default: return QString();
    }
}

QUrl buildUrl(const Account &a, const QStringList &segments)
{
    QString base = a.serverUrl.trimmed();
    while (base.endsWith(QLatin1Char('/')))
        base.chop(1);
    // Only variable segment is a 32-hex digest — no percent-encoding needed.
    return QUrl(base + QLatin1Char('/') + segments.join(QLatin1Char('/')));
}

QNetworkRequest request(const QUrl &url)
{
    QNetworkRequest req(url);
    req.setRawHeader("accept", kAccept);  // reference server 412s without it
    req.setTransferTimeout(15000);  // bound half-open sockets; times out as httpCode -1
    return req;
}

void addAuth(QNetworkRequest &req, const Account &a)
{
    req.setRawHeader("x-auth-user", a.username.toUtf8());
    req.setRawHeader("x-auth-key", a.userKey.toLower().toUtf8());  // md5 hex must be lowercase
}

Raw finish(QNetworkReply *reply)
{
    QEventLoop loop;
    QObject::connect(reply, &QNetworkReply::finished, &loop, &QEventLoop::quit);
    loop.exec();
    Raw raw;
    const QVariant code = reply->attribute(QNetworkRequest::HttpStatusCodeAttribute);
    raw.body = reply->readAll();
    if (code.isValid())
        raw.httpCode = code.toInt();
    else
        raw.transportError = reply->errorString();
    reply->deleteLater();
    return raw;
}

// Build an error Result from a non-2xx response (or transport failure).
Result errorFrom(const Raw &raw)
{
    Result r;
    r.ok = false;
    r.httpCode = raw.httpCode;
    if (raw.httpCode < 0) {
        r.message = raw.transportError.isEmpty() ? QStringLiteral("Network error")
                                                 : raw.transportError;
        return r;
    }
    const QJsonObject obj = QJsonDocument::fromJson(raw.body).object();
    const QString friendly = obj.contains(QStringLiteral("code"))
        ? friendlyMessage(obj.value(QStringLiteral("code")).toInt())
        : QString();
    if (!friendly.isEmpty())
        r.message = friendly;
    else if (obj.contains(QStringLiteral("message")))
        r.message = obj.value(QStringLiteral("message")).toString();
    else
        r.message = QStringLiteral("HTTP %1").arg(raw.httpCode);
    return r;
}

QByteArray jsonBody(const QJsonObject &obj)
{
    return QJsonDocument(obj).toJson(QJsonDocument::Compact);
}

}  // namespace

KosyncClient::KosyncClient(QNetworkAccessManager *nam) : m_nam(nam) {}

Result KosyncClient::registerUser(const Account &a)
{
    QNetworkRequest req = request(buildUrl(a, {QStringLiteral("users"), QStringLiteral("create")}));
    req.setHeader(QNetworkRequest::ContentTypeHeader, QStringLiteral("application/json"));
    QJsonObject body{{QStringLiteral("username"), a.username},
                     {QStringLiteral("password"), a.userKey}};  // register carries no auth headers
    const Raw raw = finish(m_nam->post(req, jsonBody(body)));
    if (raw.httpCode == 201)
        return Result{true, 201, QString(), {}};
    return errorFrom(raw);
}

Result KosyncClient::authorize(const Account &a)
{
    QNetworkRequest req = request(buildUrl(a, {QStringLiteral("users"), QStringLiteral("auth")}));
    addAuth(req, a);
    const Raw raw = finish(m_nam->get(req));
    if (raw.httpCode == 200)
        return Result{true, 200, QString(), {}};
    return errorFrom(raw);
}

Result KosyncClient::getProgress(const Account &a, const QString &document)
{
    QNetworkRequest req = request(
        buildUrl(a, {QStringLiteral("syncs"), QStringLiteral("progress"), document}));
    addAuth(req, a);
    const Raw raw = finish(m_nam->get(req));
    if (raw.httpCode != 200)
        return errorFrom(raw);

    const QJsonObject obj = QJsonDocument::fromJson(raw.body).object();
    Result r;
    r.ok = true;
    r.httpCode = 200;
    // "No record" is a 200 with an empty object: detected by the absent
    // percentage field, never by status.
    if (!obj.contains(QStringLiteral("percentage"))) {
        r.remote.exists = false;
        return r;
    }
    RemoteProgress &p = r.remote;
    p.exists = true;
    p.document = obj.value(QStringLiteral("document")).toString(document);
    p.progress = obj.value(QStringLiteral("progress")).toString();
    p.percentage = obj.value(QStringLiteral("percentage")).toDouble();
    p.device = obj.value(QStringLiteral("device")).toString();
    p.deviceId = obj.value(QStringLiteral("device_id")).toString();
    p.timestamp = obj.contains(QStringLiteral("timestamp"))
        ? qint64(obj.value(QStringLiteral("timestamp")).toDouble()) : -1;
    return r;
}

Result KosyncClient::updateProgress(const Account &a, const RemoteProgress &p)
{
    QNetworkRequest req = request(
        buildUrl(a, {QStringLiteral("syncs"), QStringLiteral("progress")}));
    addAuth(req, a);
    req.setHeader(QNetworkRequest::ContentTypeHeader, QStringLiteral("application/json"));
    QJsonObject body{
        {QStringLiteral("document"), p.document},
        {QStringLiteral("progress"), p.progress},
        {QStringLiteral("percentage"), truncatePercentage(p.percentage)},  // truncate, never round up
    };
    if (!p.device.isEmpty())
        body.insert(QStringLiteral("device"), p.device);
    if (!p.deviceId.isEmpty())
        body.insert(QStringLiteral("device_id"), p.deviceId);
    const Raw raw = finish(m_nam->put(req, jsonBody(body)));
    if (raw.httpCode == 200)
        return Result{true, 200, QString(), {}};
    return errorFrom(raw);
}

}  // namespace ccx::kosync
